src/common/hist_util.h

```diff
@@ -196,20 +196,10 @@ class CutsBuilder {
   }
 
   void AddCutPoint(WXQSketch::SummaryContainer const& summary) {
-    if (summary.size > 1 && summary.size <= 16) {
-      /* specialized code categorial / ordinal data -- use midpoints */
-      for (size_t i = 1; i < summary.size; ++i) {
-        bst_float cpt = (summary.data[i].value + summary.data[i - 1].value) / 2.0f;
-        if (i == 1 || cpt > p_cuts_->cut_values_.back()) {
-          p_cuts_->cut_values_.push_back(cpt);
-        }
-      }
-    } else {
-      for (size_t i = 2; i < summary.size; ++i) {
-        bst_float cpt = summary.data[i - 1].value;
-        if (i == 2 || cpt > p_cuts_->cut_values_.back()) {
-          p_cuts_->cut_values_.push_back(cpt);
-        }
+    for (size_t i = 1; i < summary.size; ++i) {
```

Comment:
@hcho3 I vaguely remembered that you have some use cases for this specialization.

Reply:
I'm still working through a bunch of corner cases for this.

Reply:
@RAMitchell My preference is either we don't support it, or provide full blown implementation.

Reply:
https://github.com/dmlc/xgboost/issues/5095 and https://github.com/dmlc/xgboost/issues/5096 may be relevant.

Reply:
The specialisation is unnecessary, we can use the same code to get correct results for both categorical and continuous. After this PR, if the number of unique feature values is less than or equal to `max_bin`, then each value will get its own bin and can be used for splitting. Before this PR, if the number of unique values was > 16 but less than `max_bin`, categorical features would get lost.